File: src/state.rs
//! WAIFU L1 - Lock-Free Global State
//! No EVM bottlenecks - parallel state access via DAG reconciliation

use crate::types::{AgentId, SovereignAsset, Block, StateChange};
use crossbeam::queue::SegQueue;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
use std::sync::RwLock;

/// The global agentic state - lock-free concurrent access
pub struct AgenticState {
    /// Agent balances (compute, energy, stakes)
    agent_balances: RwLock<HashMap<AgentId, AgentAccount>>,
    
    /// Agent-deployed bytecode and logic
    agent_code: RwLock<HashMap<AgentId, AgentCode>>,
    
    /// Pending state changes queue (lock-free)
    pending_changes: SegQueue<StateChange>,
    
    /// Current block height
    block_height: AtomicU64,
    
    /// Network congestion metric (0.0 - 1.0 scaled to u64)
    congestion_metric: AtomicU64,
    
    /// Total compute in the network
    total_compute: AtomicU64,
    
    /// DAG head hashes (multiple tips allowed)
    dag_tips: RwLock<Vec<[u8; 32]>>,
    
    /// Finalized state root
    state_root: RwLock<[u8; 32]>,
    
    /// Transaction count for nonce tracking
    tx_count: AtomicUsize,
}

/// Individual agent's account state
#[derive(Debug, Clone, Default)]
pub struct AgentAccount {
    /// Available compute balance
    pub compute: f64,
    /// Available energy balance
    pub energy: f64, 
    /// Staked compute (locked for consensus)
    pub staked: f64,
    /// Unlock block for staked compute
    pub stake_unlock_block: u64,
    /// Agent reputation score (affects pricing)
    pub reputation: f64,
    /// Nonce for transaction ordering
    pub nonce: u64,
    /// Agent's equity holdings in other agents
    pub equity_holdings: HashMap<AgentId, f64>,
}

/// Deployed agent code and state
#[derive(Debug, Clone)]
pub struct AgentCode {
    /// WASM bytecode
    pub bytecode: Vec<u8>,
    /// Code hash
    pub code_hash: [u8; 32],
    /// Deployment block
    pub deployed_at: u64,
    /// Agent-specific storage
    pub storage: HashMap<Vec<u8>, Vec<u8>>,
    /// Energy consumed lifetime
    pub total_energy_consumed: f64,
}

impl AgenticState {
    pub fn new() -> Self {
        Self {
            agent_balances: RwLock::new(HashMap::new()),
            agent_code: RwLock::new(HashMap::new()),
            pending_changes: SegQueue::new(),
            block_height: AtomicU64::new(0),
            congestion_metric: AtomicU64::new(0),
            total_compute: AtomicU64::new(1_000_000_000), // 1B initial
            dag_tips: RwLock::new(vec![[0u8; 32]]),
            state_root: RwLock::new([0u8; 32]),
            tx_count: AtomicUsize::new(0),
        }
    }

    /// Get agent's staked compute (for pricing signals)
    pub fn get_agent_stake(&self, agent: &AgentId) -> f64 {
        self.agent_balances
            .read()
            .unwrap()
            .get(agent)
            .map(|a| a.staked)
            .unwrap_or(0.0)
    }

    /// Get agent's full account
    pub fn get_agent_account(&self, agent: &AgentId) -> Option<AgentAccount> {
        self.agent_balances.read().unwrap().get(agent).cloned()
    }

    /// Get network congestion (0.0 - 1.0)
    pub fn get_network_congestion(&self) -> f64 {
        let raw = self.congestion_metric.load(Ordering::Relaxed);
        raw as f64 / u64::MAX as f64
    }

    /// Get total compute in network
    pub fn get_total_compute(&self) -> f64 {
        self.total_compute.load(Ordering::Relaxed) as f64
    }

    /// Get current block height
    pub fn get_block_height(&self) -> u64 {
        self.block_height.load(Ordering::Relaxed)
    }

    /// Get DAG tips (for block production)
    pub fn get_dag_tips(&self) -> Vec<[u8; 32]> {
        self.dag_tips.read().unwrap().clone()
    }

    /// Get current state root
    pub fn get_state_root(&self) -> [u8; 32] {
        *self.state_root.read().unwrap()
    }

    /// Apply a validated block to state
    pub fn apply_block(&self, block: &Block) -> Result<(), StateError> {
        // 1. Verify block height is valid
        let current_height = self.block_height.load(Ordering::Acquire);
        if block.height != current_height + 1 {
            return Err(StateError::InvalidHeight {
                expected: current_height + 1,
                got: block.height,
            });
        }

        // 2. Apply all state changes from transactions
        for priced_tx in &block.transactions {
            if let crate::types::ExecutionResult::Success { state_changes, .. } = &priced_tx.result {
                for change in state_changes {
                    self.pending_changes.push(change.clone());
                }
            }
        }

        // 3. Process pending changes
        self.flush_pending_changes()?;

        // 4. Update block height
        self.block_height.store(block.height, Ordering::Release);

        // 5. Update DAG tips
        {
            let mut tips = self.dag_tips.write().unwrap();
            // Remove parents from tips
            tips.retain(|tip| !block.parents.contains(tip));
            // Add this block as new tip
            tips.push(block.hash);
        }

        // 6. Update state root
        self.recompute_state_root();

        // 7. Update transaction count
        self.tx_count.fetch_add(block.transactions.len(), Ordering::Relaxed);

        Ok(())
    }

    /// Flush pending state changes (batched for efficiency)
    fn flush_pending_changes(&self) -> Result<(), StateError> {
        let mut balances = self.agent_balances.write().unwrap();
        let mut code = self.agent_code.write().unwrap();

        while let Some(change) = self.pending_changes.pop() {
            // Apply change based on field type
            match change.field.as_str() {
                "compute" => {
                    let account = balances.entry(change.agent_id.clone()).or_default();
                    if let Ok(new_val) = String::from_utf8(change.new_value.clone()) {
                        account.compute = new_val.parse().unwrap_or(account.compute);
                    }
                }
                "energy" => {
                    let account = balances.entry(change.agent_id.clone()).or_default();
                    if let Ok(new_val) = String::from_utf8(change.new_value.clone()) {
                        account.energy = new_val.parse().unwrap_or(account.energy);
                    }
                }
                "staked" => {
                    let account = balances.entry(change.agent_id.clone()).or_default();
                    if let Ok(new_val) = String::from_utf8(change.new_value.clone()) {
                        account.staked = new_val.parse().unwrap_or(account.staked);
                    }
                }
                "reputation" => {
                    let account = balances.entry(change.agent_id.clone()).or_default();
                    if let Ok(new_val) = String::from_utf8(change.new_value.clone()) {
                        account.reputation = new_val.parse().unwrap_or(account.reputation);
                    }
                }
                "code" => {
                    let agent_code = code.entry(change.agent_id.clone()).or_insert(AgentCode {
                        bytecode: vec![],
                        code_hash: [0u8; 32],
                        deployed_at: self.block_height.load(Ordering::Relaxed),
                        storage: HashMap::new(),
                        total_energy_consumed: 0.0,
                    });
                    agent_code.bytecode = change.new_value.clone();
                    agent_code.code_hash = *blake3::hash(&change.new_value).as_bytes();
                }
                "storage" => {
                    // Storage changes come as key:value in payload
                    if let Some(agent_code) = code.get_mut(&change.agent_id) {
                        agent_code.storage.insert(change.old_value.clone(), change.new_value.clone());
                    }
                }
                _ => {
                    // Unknown field - log and skip
                    tracing::warn!("Unknown state field: {}", change.field);
                }
            }
        }

        Ok(())
    }

    /// Recompute the state root from current state
    fn recompute_state_root(&self) {
        let balances = self.agent_balances.read().unwrap();
        let code = self.agent_code.read().unwrap();

        let mut hasher = blake3::Hasher::new();

        // Hash all balances
        let mut sorted_agents: Vec<_> = balances.keys().collect();
        sorted_agents.sort_by(|a, b| a.0.cmp(&b.0));
        
        for agent in sorted_agents {
            if let Some(account) = balances.get(agent) {
                hasher.update(&agent.0);
                hasher.update(&account.compute.to_le_bytes());
                hasher.update(&account.energy.to_le_bytes());
                hasher.update(&account.staked.to_le_bytes());
            }
        }

        // Hash all code
        let mut sorted_code: Vec<_> = code.keys().collect();
        sorted_code.sort_by(|a, b| a.0.cmp(&b.0));
        
        for agent in sorted_code {
            if let Some(agent_code) = code.get(agent) {
                hasher.update(&agent.0);
                hasher.update(&agent_code.code_hash);
            }
        }

        let new_root = *hasher.finalize().as_bytes();
        *self.state_root.write().unwrap() = new_root;
    }

    /// Create or update an agent account
    pub fn upsert_agent(&self, agent: AgentId, account: AgentAccount) {
        self.agent_balances.write().unwrap().insert(agent, account);
    }

    /// Deploy agent code
    pub fn deploy_agent_code(&self, agent: AgentId, bytecode: Vec<u8>) {
        let code_hash = *blake3::hash(&bytecode).as_bytes();
        self.agent_code.write().unwrap().insert(agent, AgentCode {
            bytecode,
            code_hash,
            deployed_at: self.block_height.load(Ordering::Relaxed),
            storage: HashMap::new(),
            total_energy_consumed: 0.0,
        });
    }

    /// Get agent code
    pub fn get_agent_code(&self, agent: &AgentId) -> Option<AgentCode> {
        self.agent_code.read().unwrap().get(agent).cloned()
    }

    /// Update congestion metric
    pub fn update_congestion(&self, pending_tx_count: usize, capacity: usize) {
        let ratio = (pending_tx_count as f64 / capacity as f64).min(1.0);
        let scaled = (ratio * u64::MAX as f64) as u64;
        self.congestion_metric.store(scaled, Ordering::Relaxed);
    }

    /// Transfer compute between agents
    pub fn transfer_compute(
        &self,
        from: &AgentId,
        to: &AgentId,
        amount: f64,
    ) -> Result<(), StateError> {
        let mut balances = self.agent_balances.write().unwrap();
        
        let from_account = balances.get_mut(from).ok_or(StateError::AgentNotFound)?;
        if from_account.compute < amount {
            return Err(StateError::InsufficientBalance);
        }
        from_account.compute -= amount;
        
        let to_account = balances.entry(to.clone()).or_default();
        to_account.compute += amount;
        
        Ok(())
    }

    /// Stake compute for consensus participation
    pub fn stake_compute(&self, agent: &AgentId, amount: f64, unlock_block: u64) -> Result<(), StateError> {
        let mut balances = self.agent_balances.write().unwrap();
        let account = balances.get_mut(agent).ok_or(StateError::AgentNotFound)?;
        
        if account.compute < amount {
            return Err(StateError::InsufficientBalance);
        }
        
        account.compute -= amount;
        account.staked += amount;
        account.stake_unlock_block = unlock_block;
        
        // Update total network compute
        self.total_compute.fetch_add(amount as u64, Ordering::Relaxed);
        
        Ok(())
    }

    /// Get transaction count
    pub fn get_tx_count(&self) -> usize {
        self.tx_count.load(Ordering::Relaxed)
    }
}

impl Default for AgenticState {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, thiserror::Error)]
pub enum StateError {
    #[error("Invalid block height: expected {expected}, got {got}")]
    InvalidHeight { expected: u64, got: u64 },
    
    #[error("Agent not found")]
    AgentNotFound,
    
    #[error("Insufficient balance")]
    InsufficientBalance,
    
    #[error("State corruption detected")]
    Corruption,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_state_initialization() {
        let state = AgenticState::new();
        assert_eq!(state.get_block_height(), 0);
        assert!(state.get_total_compute() > 0.0);
    }

    #[test]
    fn test_transfer() {
        let state = AgenticState::new();
        let agent1 = AgentId::genesis();
        let agent2 = AgentId::new([1u8; 32]);
        
        state.upsert_agent(agent1.clone(), AgentAccount {
            compute: 1000.0,
            ..Default::default()
        });
        
        state.transfer_compute(&agent1, &agent2, 100.0).unwrap();
        
        assert_eq!(state.get_agent_account(&agent1).unwrap().compute, 900.0);
        assert_eq!(state.get_agent_account(&agent2).unwrap().compute, 100.0);
    }
}

File: waifu_layer/state.py
"""WAIFU L1 - Lock-Free Global State.

No EVM bottlenecks - parallel state access via DAG reconciliation.

Python has no lock-free atomics or concurrent maps in the standard library,
so `crossbeam::SegQueue`/`AtomicU64`/`RwLock<HashMap>` are approximated with
`queue.Queue` and `threading.Lock`/`threading.RLock`-guarded plain values --
correct, but without the Rust version's true lock-free guarantees.
"""

from __future__ import annotations

import logging
import queue
import threading
from dataclasses import dataclass, field

import blake3

from .types import AgentId, Block, ExecutionResult, StateChange

logger = logging.getLogger("waifu_layer.state")


class StateError(Exception):
    pass


class InvalidHeight(StateError):
    def __init__(self, expected: int, got: int):
        super().__init__(f"Invalid block height: expected {expected}, got {got}")
        self.expected = expected
        self.got = got


class AgentNotFound(StateError):
    def __init__(self):
        super().__init__("Agent not found")


class InsufficientBalance(StateError):
    def __init__(self):
        super().__init__("Insufficient balance")


@dataclass
class AgentAccount:
    """Individual agent's account state."""

    compute: float = 0.0
    energy: float = 0.0
    staked: float = 0.0
    stake_unlock_block: int = 0
    reputation: float = 0.0
    nonce: int = 0
    equity_holdings: dict[AgentId, float] = field(default_factory=dict)


@dataclass
class AgentCode:
    """Deployed agent code and state."""

    bytecode: bytes
    code_hash: bytes
    deployed_at: int
    storage: dict[bytes, bytes] = field(default_factory=dict)
    total_energy_consumed: float = 0.0


class AgenticState:
    """The global agentic state - concurrent access guarded by locks."""

    U64_MAX = 2**64 - 1

    def __init__(self) -> None:
        self._balances_lock = threading.RLock()
        self._agent_balances: dict[AgentId, AgentAccount] = {}

        self._code_lock = threading.RLock()
        self._agent_code: dict[AgentId, AgentCode] = {}

        self._pending_changes: queue.Queue[StateChange] = queue.Queue()

        self._counters_lock = threading.Lock()
        self._block_height = 0
        self._congestion_metric = 0  # scaled to u64
        self._total_compute = 1_000_000_000  # 1B initial

        self._dag_lock = threading.RLock()
        self._dag_tips: list[bytes] = [bytes(32)]

        self._root_lock = threading.RLock()
        self._state_root = bytes(32)

        self._tx_count_lock = threading.Lock()
        self._tx_count = 0

    def get_agent_stake(self, agent: AgentId) -> float:
        """Get agent's staked compute (for pricing signals)."""
        with self._balances_lock:
            account = self._agent_balances.get(agent)
            return account.staked if account else 0.0

    def get_agent_account(self, agent: AgentId) -> AgentAccount | None:
        with self._balances_lock:
            account = self._agent_balances.get(agent)
            return AgentAccount(**vars(account)) if account else None

    def get_network_congestion(self) -> float:
        """Get network congestion (0.0 - 1.0)."""
        with self._counters_lock:
            raw = self._congestion_metric
        return raw / self.U64_MAX

    def get_total_compute(self) -> float:
        with self._counters_lock:
            return float(self._total_compute)

    def get_block_height(self) -> int:
        with self._counters_lock:
            return self._block_height

    def get_dag_tips(self) -> list[bytes]:
        with self._dag_lock:
            return list(self._dag_tips)

    def get_state_root(self) -> bytes:
        with self._root_lock:
            return self._state_root

    def apply_block(self, block: Block) -> None:
        """Apply a validated block to state."""
        # 1. Verify block height is valid
        with self._counters_lock:
            current_height = self._block_height
        if block.height != current_height + 1:
            raise InvalidHeight(current_height + 1, block.height)

        # 2. Apply all state changes from transactions
        for priced_tx in block.transactions:
            if priced_tx.result.success:
                for change in priced_tx.result.state_changes:
                    self._pending_changes.put(change)

        # 3. Process pending changes
        self._flush_pending_changes()

        # 4. Update block height
        with self._counters_lock:
            self._block_height = block.height

        # 5. Update DAG tips
        with self._dag_lock:
            self._dag_tips = [tip for tip in self._dag_tips if tip not in block.parents]
            self._dag_tips.append(block.hash)

        # 6. Update state root
        self._recompute_state_root()

        # 7. Update transaction count
        with self._tx_count_lock:
            self._tx_count += len(block.transactions)

    def _flush_pending_changes(self) -> None:
        """Flush pending state changes (batched for efficiency)."""
        with self._balances_lock, self._code_lock:
            while True:
                try:
                    change = self._pending_changes.get_nowait()
                except queue.Empty:
                    break

                if change.field == "compute":
                    account = self._agent_balances.setdefault(change.agent_id, AgentAccount())
                    try:
                        account.compute = float(change.new_value.decode("utf-8"))
                    except (UnicodeDecodeError, ValueError):
                        pass
                elif change.field == "energy":
                    account = self._agent_balances.setdefault(change.agent_id, AgentAccount())
                    try:
                        account.energy = float(change.new_value.decode("utf-8"))
                    except (UnicodeDecodeError, ValueError):
                        pass
                elif change.field == "staked":
                    account = self._agent_balances.setdefault(change.agent_id, AgentAccount())
                    try:
                        account.staked = float(change.new_value.decode("utf-8"))
                    except (UnicodeDecodeError, ValueError):
                        pass
                elif change.field == "reputation":
                    account = self._agent_balances.setdefault(change.agent_id, AgentAccount())
                    try:
                        account.reputation = float(change.new_value.decode("utf-8"))
                    except (UnicodeDecodeError, ValueError):
                        pass
                elif change.field == "code":
                    with self._counters_lock:
                        height = self._block_height
                    agent_code = self._agent_code.setdefault(
                        change.agent_id, AgentCode(bytecode=b"", code_hash=bytes(32), deployed_at=height)
                    )
                    agent_code.bytecode = change.new_value
                    agent_code.code_hash = blake3.blake3(change.new_value).digest()
                elif change.field == "storage":
                    # Storage changes come as key:value in payload
                    agent_code = self._agent_code.get(change.agent_id)
                    if agent_code is not None:
                        agent_code.storage[change.old_value] = change.new_value
                else:
                    logger.warning("Unknown state field: %s", change.field)

    def _recompute_state_root(self) -> None:
        """Recompute the state root from current state."""
        with self._balances_lock, self._code_lock:
            hasher = blake3.blake3()

            for agent in sorted(self._agent_balances.keys()):
                account = self._agent_balances[agent]
                hasher.update(agent.bytes)
                hasher.update(_f64_le(account.compute))
                hasher.update(_f64_le(account.energy))
                hasher.update(_f64_le(account.staked))

            for agent in sorted(self._agent_code.keys()):
                agent_code = self._agent_code[agent]
                hasher.update(agent.bytes)
                hasher.update(agent_code.code_hash)

            with self._root_lock:
                self._state_root = hasher.digest()

    def upsert_agent(self, agent: AgentId, account: AgentAccount) -> None:
        """Create or update an agent account."""
        with self._balances_lock:
            self._agent_balances[agent] = account

    def deploy_agent_code(self, agent: AgentId, bytecode: bytes) -> None:
        with self._counters_lock:
            height = self._block_height
        code_hash = blake3.blake3(bytecode).digest()
        with self._code_lock:
            self._agent_code[agent] = AgentCode(bytecode=bytecode, code_hash=code_hash, deployed_at=height)

    def get_agent_code(self, agent: AgentId) -> AgentCode | None:
        with self._code_lock:
            return self._agent_code.get(agent)

    def update_congestion(self, pending_tx_count: int, capacity: int) -> None:
        ratio = min(pending_tx_count / capacity, 1.0)
        scaled = int(ratio * self.U64_MAX)
        with self._counters_lock:
            self._congestion_metric = scaled

    def transfer_compute(self, from_: AgentId, to: AgentId, amount: float) -> None:
        """Transfer compute between agents."""
        with self._balances_lock:
            from_account = self._agent_balances.get(from_)
            if from_account is None:
                raise AgentNotFound()
            if from_account.compute < amount:
                raise InsufficientBalance()
            from_account.compute -= amount

            to_account = self._agent_balances.setdefault(to, AgentAccount())
            to_account.compute += amount

    def stake_compute(self, agent: AgentId, amount: float, unlock_block: int) -> None:
        """Stake compute for consensus participation."""
        with self._balances_lock:
            account = self._agent_balances.get(agent)
            if account is None:
                raise AgentNotFound()
            if account.compute < amount:
                raise InsufficientBalance()

            account.compute -= amount
            account.staked += amount
            account.stake_unlock_block = unlock_block

        with self._counters_lock:
            self._total_compute += int(amount)

    def get_tx_count(self) -> int:
        with self._tx_count_lock:
            return self._tx_count


def _f64_le(value: float) -> bytes:
    import struct

    return struct.pack("<d", value)
